File: python/periph/chips/adc_dac/mcp4728.py
class MCP4728Minimal:
    """MCP4728 quad-channel 12-bit voltage-output DAC — minimal interface.

    Provides simple voltage output as a fraction of V_DD for any of the four
    channels (A–D) plus a convenience method to update all four channels
    simultaneously. No configuration required beyond the transport. V_REF is
    fixed at external (V_DD), gain is fixed at ×1, and power-down is off.
    EEPROM is never written by this class.

    Args:
        transport: Configured I²C transport pointing at the device (0x60–0x67).
    """

    _CMD_MULTI_WRITE = 0x40  # C2=0 C1=0 C0=0 — see spec; Multi-Write is 010
    # Multi-Write byte 1: [0 1 0 0 0 DAC1 DAC0 UDAC] = 0x40 | (channel << 1)
    _CMD_FAST_WRITE_BASE = 0x00  # Per-channel fast write: 2 bytes each, A→D

    def __init__(self, transport):
        """Initialize MCP4728Minimal and store the transport.

        Args:
            transport: Configured I²C transport pointing at the device.
        """
        self._transport = transport

    def set_voltage(self, channel, fraction):
        """Set one channel's DAC output as a fraction of V_DD.

        Uses Multi-Write to update a single channel's volatile DAC register.
        V_REF=0 (external V_DD), gain=1, PD=00, UDAC=0 (immediate V_OUT
        update). EEPROM is not written.

        Args:
            channel: Channel index 0 (A) – 3 (D).
            fraction: Output voltage as a fraction of V_DD (0.0–1.0).
        """
        code = int(round(max(0.0, min(1.0, fraction)) * 4095))
        self.set_raw(channel, code)

    def set_raw(self, channel, code):
        """Set one channel's raw 12-bit DAC code.

        Uses Multi-Write. V_REF=0 (external V_DD), gain=1, PD=00, UDAC=0.
        EEPROM is not written.

        Args:
            channel: Channel index 0 (A) – 3 (D).
            code: Raw 12-bit DAC code (0–4095).
        """
        ch = max(0, min(3, channel))
        c = int(max(0, min(4095, code)))
        self._multi_write(ch, c, vref=0, pd=0, gain=0, udac=0)

    def set_all(self, fractions):
        """Update all four channels simultaneously using Fast Write.

        Issues a single 8-byte Fast Write transaction that updates channels
        A→D at once. V_REF and gain bits are not carried by Fast Write; they
        retain whatever values are currently in each channel's input
        register. PD=00, UDAC=0 for all four channels. EEPROM is not written.

        Args:
            fractions: Iterable of 4 fractions (0.0–1.0), index 0 = A.
        """
        if len(fractions) != 4:
            raise ValueError("fractions must have exactly 4 elements")
        out = bytearray(8)
        for i in range(4):
            code = int(round(max(0.0, min(1.0, fractions[i])) * 4095))
            code = max(0, min(4095, code))
            # First channel: byte1 = [0 0 PD1 PD0 D11-D8]; channels B-D: same
            out[i * 2] = ((code >> 8) & 0x0F)  # PD1=PD0=0
            out[i * 2 + 1] = code & 0xFF
        self._transport.write(bytes(out))

    def _multi_write(self, channel, code, vref, pd, gain, udac):
        byte1 = 0x40 | ((channel & 0x03) << 1) | (udac & 0x01)
        byte2 = ((vref & 0x01) << 7) | ((pd & 0x03) << 5) | \
                ((gain & 0x01) << 4) | ((code >> 8) & 0x0F)
        byte3 = code & 0xFF
        self._transport.write(bytes([byte1, byte2, byte3]))


class MCP4728Full(MCP4728Minimal):
    """MCP4728 full interface — extends MCP4728Minimal with EEPROM, V_REF, gain, power-down, and read-back.

    Adds per-channel V_REF and gain configuration, all-channel V_REF/gain/
    power-down commands, write-with-EEPROM persistence (Single and Sequential
    Write), General Call reset/wake-up/software-update, and full 24-byte
    read-back of all channel DAC input registers and EEPROM contents.

    Args:
        transport: Configured I²C transport pointing at the device (0x60–0x67).
    """

    _CMD_SINGLE_WRITE    = 0x58  # [0 1 0 1 1 DAC1 DAC0 UDAC]
    _CMD_SEQUENTIAL_BASE = 0x50  # [0 1 0 1 0 DAC1 DAC0 UDAC] (start channel)
    _CMD_WRITE_VREF      = 0x80  # [1 0 0 X Vref_A Vref_B Vref_C Vref_D]
    _CMD_WRITE_GAIN      = 0xC0  # [1 1 0 X Gx_A Gx_B Gx_C Gx_D]
    _CMD_WRITE_POWERDOWN = 0xA0  # [1 0 1 X ...]
    _ADDR_GENERAL_CALL   = 0x00

    _GC_RESET        = 0x06
    _GC_SOFTWARE_UPD = 0x08
    _GC_WAKE         = 0x09

    PD_NORMAL    = 0
    PD_1K_GND    = 1
    PD_100K_GND  = 2
    PD_500K_GND  = 3

    VREF_EXTERNAL = 0  # V_DD
    VREF_INTERNAL = 1  # 2.048 V

    GAIN_X1 = 0
    GAIN_X2 = 1

    def __init__(self, transport):
        """Initialize MCP4728Full and store the transport.

        Args:
            transport: Configured I²C transport pointing at the device.
        """
        super().__init__(transport)

    def set_voltage_eeprom(self, channel, fraction, vref, gain):
        """Set one channel's output and persist to EEPROM.

        Uses Single Write: updates the volatile DAC input register and the
        nonvolatile EEPROM copy. Output voltage is computed from the
        channel's configured V_REF and gain.

        Args:
            channel: Channel index 0 (A) – 3 (D).
            fraction: Output as a fraction of the configured full-scale
                     (V_DD for external, 2.048 V for internal ×1, 4.096 V for
                     internal ×2).
            vref: 0 = external (V_DD), 1 = internal (2.048 V).
            gain: 1 = ×1, 2 = ×2 (ignored when vref = external).
        """
        f = max(0.0, min(1.0, float(fraction)))
        code = int(round(f * 4095))
        self._single_write(channel, code, vref, pd=0, gain=gain, udac=0)

    def set_raw_eeprom(self, channel, code, vref, gain):
        """Set one channel's raw 12-bit code and persist to EEPROM.

        Args:
            channel: Channel index 0 (A) – 3 (D).
            code: Raw 12-bit DAC code (0–4095).
            vref: 0 = external (V_DD), 1 = internal (2.048 V).
            gain: 1 = ×1, 2 = ×2 (ignored when vref = external).
        """
        c = int(max(0, min(4095, code)))
        self._single_write(channel, c, vref, pd=0, gain=gain, udac=0)

    def set_all_eeprom(self, fractions, vrefs, gains):
        """Update all four channels and EEPROM starting from channel A.

        Uses Sequential Write from channel 0 to channel 3. Each channel
        receives one 2-byte (V_REF/PD/Gain/code-high, code-low) pair; the
        chip writes all four DAC registers and persists the EEPROM at the
        end of the transaction (8 data bytes plus the 1 command byte).

        Args:
            fractions: Iterable of 4 fractions (0.0–1.0), index 0 = A.
            vrefs: Iterable of 4 V_REF values (0/1).
            gains: Iterable of 4 gain values (1/2).
        """
        if len(fractions) != 4 or len(vrefs) != 4 or len(gains) != 4:
            raise ValueError("fractions, vrefs, gains must each have 4 elements")
        out = bytearray(9)
        # Command byte: Sequential Write starting at channel 0
        out[0] = self._CMD_SEQUENTIAL_BASE | 0x00  # DAC1=DAC0=0 (A), UDAC=0
        for i in range(4):
            f = max(0.0, min(1.0, float(fractions[i])))
            code = int(round(f * 4095))
            if code > 4095:
                code = 4095
            v = 1 if vrefs[i] else 0
            g = 1 if gains[i] == 2 else 0
            # Per-channel byte layout (Multi-Write format): [V_REF PD1 PD0 Gx D11-D8]
            out[1 + i * 2] = ((v & 0x01) << 7) | ((g & 0x01) << 4) | ((code >> 8) & 0x0F)
            out[1 + i * 2 + 1] = code & 0xFF
        self._transport.write(bytes(out))

    def set_vref(self, vref_a, vref_b, vref_c, vref_d):
        """Set V_REF for all four channels (volatile register only).

        Args:
            vref_a: VREF for channel A (0 = external/V_DD, 1 = internal 2.048 V).
            vref_b: VREF for channel B.
            vref_c: VREF for channel C.
            vref_d: VREF for channel D.
        """
        byte1 = self._CMD_WRITE_VREF | (
            (self._b1(vref_a) << 3) | (self._b1(vref_b) << 2) |
            (self._b1(vref_c) << 1) | self._b1(vref_d))
        self._transport.write(bytes([byte1]))

    def set_gain(self, gain_a, gain_b, gain_c, gain_d):
        """Set gain for all four channels (volatile register only).

        Args:
            gain_a: Gain for channel A (1 = ×1, 2 = ×2).
            gain_b: Gain for channel B.
            gain_c: Gain for channel C.
            gain_d: Gain for channel D.
        """
        byte1 = self._CMD_WRITE_GAIN | (
            (self._g(gain_a) << 3) | (self._g(gain_b) << 2) |
            (self._g(gain_c) << 1) | self._g(gain_d))
        self._transport.write(bytes([byte1]))

    def set_power_down(self, pd_a, pd_b, pd_c, pd_d):
        """Set power-down mode for all four channels (volatile register only).

        Args:
            pd_a: Power-down mode for A (0–3).
            pd_b: Power-down mode for B (0–3).
            pd_c: Power-down mode for C (0–3).
            pd_d: Power-down mode for D (0–3).
        """
        byte1 = self._CMD_WRITE_POWERDOWN | (
            (self._p2(pd_a) << 4) | (self._p1(pd_a) << 3) |
            (self._p2(pd_b) << 2) | (self._p1(pd_b) << 1))
        byte2 = ((self._p2(pd_c) << 6) | (self._p1(pd_c) << 5) |
                 (self._p2(pd_d) << 4) | (self._p1(pd_d) << 3))
        self._transport.write(bytes([byte1, byte2]))

    def read(self):
        """Read all four channels' DAC input registers and EEPROM contents.

        The chip returns 24 bytes: 4 channels × (3 bytes input register +
        3 bytes EEPROM). Each 3-byte group has the same layout: byte 1 holds
        RDY/BSY + POR + channel ID + I²C address bits; bytes 2–3 hold
        V_REF, PD, gain, and the 12-bit code.

        Returns:
            list: 4 dicts, one per channel, in order A, B, C, D. Each dict
                  contains: code, vref, gain, power_down, eeprom_code,
                  eeprom_vref, eeprom_gain, eeprom_power_down, eeprom_ready.
        """
        buf = self._transport.read(24)
        result = []
        eeprom_ready = bool(buf[0] & 0x80)
        for i in range(4):
            base = i * 3
            inp_vref  = (buf[base + 1] >> 7) & 0x01
            inp_pd    = (buf[base + 1] >> 5) & 0x03
            inp_gain  = (buf[base + 1] >> 4) & 0x01
            inp_code  = ((buf[base + 1] & 0x0F) << 8) | buf[base + 2]
            result.append({
                'code': inp_code,
                'vref': inp_vref,
                'gain': 2 if inp_gain else 1,
                'power_down': inp_pd,
            })
        for i in range(4):
            base = 12 + i * 3
            ee_vref  = (buf[base + 1] >> 7) & 0x01
            ee_pd    = (buf[base + 1] >> 5) & 0x03
            ee_gain  = (buf[base + 1] >> 4) & 0x01
            ee_code  = ((buf[base + 1] & 0x0F) << 8) | buf[base + 2]
            result[i]['eeprom_code'] = ee_code
            result[i]['eeprom_vref'] = ee_vref
            result[i]['eeprom_gain'] = 2 if ee_gain else 1
            result[i]['eeprom_power_down'] = ee_pd
            result[i]['eeprom_ready'] = eeprom_ready
        return result

    def is_eeprom_ready(self):
        """Check whether an EEPROM write is in progress.

        Returns:
            bool: True when no EEPROM write is pending (RDY/BSY = 1).
        """
        buf = self._transport.read(1)
        return bool(buf[0] & 0x80)

    def software_update(self):
        """Send General Call Software Update (0x00, 0x08) to latch all V_OUT."""
        self._transport.write(bytes([self._ADDR_GENERAL_CALL, self._GC_SOFTWARE_UPD]))

    def wake_up(self):
        """Send General Call Wake-Up (0x00, 0x09) to clear all PD bits."""
        self._transport.write(bytes([self._ADDR_GENERAL_CALL, self._GC_WAKE]))

    def reset(self):
        """Send General Call Reset (0x00, 0x06) to reload EEPROM into all DAC registers."""
        self._transport.write(bytes([self._ADDR_GENERAL_CALL, self._GC_RESET]))

    def _single_write(self, channel, code, vref, pd, gain, udac):
        ch = max(0, min(3, channel))
        c = int(max(0, min(4095, code)))
        byte1 = self._CMD_SINGLE_WRITE | ((ch & 0x03) << 1) | (udac & 0x01)
        byte2 = ((vref & 0x01) << 7) | ((pd & 0x03) << 5) | \
                ((gain & 0x01) << 4) | ((c >> 8) & 0x0F)
        byte3 = c & 0xFF
        self._transport.write(bytes([byte1, byte2, byte3]))

    @staticmethod
    def _b1(v):
        return 1 if v else 0

    @staticmethod
    def _g(g):
        return 1 if g == 2 else 0

    @staticmethod
    def _p1(pd):
        return (pd >> 0) & 0x01

    @staticmethod
    def _p2(pd):
        return (pd >> 1) & 0x01

File: rust/periph/src/chips/adc_dac/mcp4728.rs
//! MCP4728 quad-channel 12-bit voltage-output DAC (Microchip).
//!
//! Communicates over I²C (up to 3.4 MHz high-speed, 400 kHz fast mode, 100 kHz standard).
//! Provides voltage output as a fraction of V_DD for any of the four channels
//! (A–D) plus a convenience method to update all four channels simultaneously.
//! V_REF is fixed at external (V_DD), gain is fixed at ×1, and power-down is
//! off in the minimal interface. The chip has on-chip EEPROM that auto-loads
//! into the DAC registers on power-up.

use embedded_hal::i2c::I2c;

const CMD_MULTI_WRITE_BASE: u8 = 0x40; // [0 1 0 0 0 DAC1 DAC0 UDAC]
const CMD_SINGLE_WRITE: u8     = 0x58; // [0 1 0 1 1 DAC1 DAC0 UDAC]
const CMD_SEQUENTIAL_BASE: u8  = 0x50; // [0 1 0 1 0 DAC1 DAC0 UDAC]
const CMD_WRITE_VREF: u8       = 0x80; // [1 0 0 X Vref_A Vref_B Vref_C Vref_D]
const CMD_WRITE_GAIN: u8       = 0xC0; // [1 1 0 X Gx_A Gx_B Gx_C Gx_D]
const CMD_WRITE_POWERDOWN: u8  = 0xA0; // [1 0 1 X ...]
const ADDR_GENERAL_CALL: u8    = 0x00;
const GC_RESET: u8             = 0x06;
const GC_SOFTWARE_UPD: u8      = 0x08;
const GC_WAKE: u8              = 0x09;

/// MCP4728 minimal driver — set voltage as fraction or raw 12-bit code on any
/// of the four channels, or update all four simultaneously.
pub struct Mcp4728Minimal<I2C> {
    i2c: I2C,
    addr: u8,
}

impl<I2C: I2c> Mcp4728Minimal<I2C> {
    /// Create a new `Mcp4728Minimal`.
    ///
    /// # Arguments
    /// * `i2c`  — Configured I²C bus implementing [`embedded_hal::i2c::I2c`].
    /// * `addr` — 7-bit device address (typically `0x60`–`0x67`).
    pub fn new(i2c: I2C, addr: u8) -> Result<Self, I2C::Error> {
        Ok(Self { i2c, addr })
    }

    /// Set one channel's DAC output as a fraction of V_DD.
    ///
    /// Clamps to [0.0, 1.0] and uses Multi-Write to update a single channel's
    /// volatile DAC register. V_REF=external, gain=×1, PD=00, UDAC=0.
    pub fn set_voltage(&mut self, channel: u8, fraction: f32) -> Result<(), I2C::Error> {
        let f = fraction.max(0.0).min(1.0);
        let code = (f * 4095.0) as u16;
        self.set_raw(channel, code)
    }

    /// Set one channel's raw 12-bit DAC code.
    ///
    /// Clamps to [0, 4095] and uses Multi-Write. EEPROM is not written.
    pub fn set_raw(&mut self, channel: u8, code: u16) -> Result<(), I2C::Error> {
        let ch = channel.min(3);
        let c = code.min(4095);
        self._multi_write(ch, c, 0, 0, 0, 0)
    }

    /// Update all four channels simultaneously using Fast Write.
    ///
    /// Issues a single 8-byte Fast Write transaction that updates channels
    /// A→D at once. PD=00, UDAC=0 for all four channels. EEPROM is not
    /// written.
    pub fn set_all(&mut self, fractions: [f32; 4]) -> Result<(), I2C::Error> {
        let mut buf = [0u8; 8];
        for i in 0..4 {
            let f = fractions[i].max(0.0).min(1.0);
            let code = ((f * 4095.0) as u16).min(4095);
            buf[i * 2]     = ((code >> 8) & 0x0F) as u8;
            buf[i * 2 + 1] = (code & 0xFF) as u8;
        }
        self.i2c.write(self.addr, &buf)
    }

    fn _multi_write(&mut self, channel: u8, code: u16, vref: u8, pd: u8, gain: u8, udac: u8) -> Result<(), I2C::Error> {
        let buf = [
            CMD_MULTI_WRITE_BASE | ((channel & 0x03) << 1) | (udac & 0x01),
            ((vref & 0x01) << 7) | ((pd & 0x03) << 5) | ((gain & 0x01) << 4) | ((code >> 8) & 0x0F) as u8,
            (code & 0xFF) as u8,
        ];
        self.i2c.write(self.addr, &buf)
    }
}

/// MCP4728 full driver — extends [`Mcp4728Minimal`] with EEPROM, V_REF, gain, power-down, and read-back.
pub struct Mcp4728Full<I2C> {
    inner: Mcp4728Minimal<I2C>,
}

/// Per-channel state read from the chip.
#[derive(Clone, Copy, Debug)]
pub struct ChannelState {
    /// DAC input register code (0–4095).
    pub code: u16,
    /// V_REF selection: 0 = external (V_DD), 1 = internal (2.048 V).
    pub vref: u8,
    /// Gain: 1 = ×1, 2 = ×2.
    pub gain: u8,
    /// Power-down mode 0–3.
    pub power_down: u8,
    /// EEPROM-stored DAC code (0–4095).
    pub eeprom_code: u16,
    /// EEPROM-stored V_REF (0/1).
    pub eeprom_vref: u8,
    /// EEPROM-stored gain (1/2).
    pub eeprom_gain: u8,
    /// EEPROM-stored power-down mode (0–3).
    pub eeprom_power_down: u8,
}

/// All four channels plus the EEPROM-ready flag.
#[derive(Clone, Copy, Debug)]
pub struct ReadResult {
    pub channel: [ChannelState; 4],
    pub eeprom_ready: bool,
}

impl<I2C: I2c> Mcp4728Full<I2C> {
    /// Create a new `Mcp4728Full`.
    ///
    /// # Arguments
    /// * `i2c`  — Configured I²C bus implementing [`embedded_hal::i2c::I2c`].
    /// * `addr` — 7-bit device address (typically `0x60`–`0x67`).
    pub fn new(i2c: I2C, addr: u8) -> Result<Self, I2C::Error> {
        Ok(Self { inner: Mcp4728Minimal::new(i2c, addr)? })
    }

    pub fn set_voltage(&mut self, channel: u8, fraction: f32) -> Result<(), I2C::Error> {
        self.inner.set_voltage(channel, fraction)
    }

    pub fn set_raw(&mut self, channel: u8, code: u16) -> Result<(), I2C::Error> {
        self.inner.set_raw(channel, code)
    }

    pub fn set_all(&mut self, fractions: [f32; 4]) -> Result<(), I2C::Error> {
        self.inner.set_all(fractions)
    }

    /// Set one channel's output and persist to EEPROM (Single Write).
    pub fn set_voltage_eeprom(&mut self, channel: u8, fraction: f32, vref: u8, gain: u8) -> Result<(), I2C::Error> {
        let f = fraction.max(0.0).min(1.0);
        let code = (f * 4095.0) as u16;
        self._single_write(channel, code, vref, 0, gain, 0)
    }

    /// Set one channel's raw 12-bit code and persist to EEPROM.
    pub fn set_raw_eeprom(&mut self, channel: u8, code: u16, vref: u8, gain: u8) -> Result<(), I2C::Error> {
        let ch = channel.min(3);
        let c = code.min(4095);
        self._single_write(ch, c, vref, 0, gain, 0)
    }

    /// Update all four channels and EEPROM (Sequential Write from A to D).
    pub fn set_all_eeprom(&mut self, fractions: [f32; 4], vrefs: [u8; 4], gains: [u8; 4]) -> Result<(), I2C::Error> {
        let mut buf = [0u8; 9];
        buf[0] = CMD_SEQUENTIAL_BASE | 0x00;
        for i in 0..4 {
            let f = fractions[i].max(0.0).min(1.0);
            let code = ((f * 4095.0) as u16).min(4095);
            let v = vrefs[i] & 0x01;
            let g = if gains[i] == 2 { 1 } else { 0 };
            // Per-channel byte layout (Multi-Write format): [V_REF PD1 PD0 Gx D11-D8]
            // PD bits are always 0 here (no power-down is set in this command).
            buf[1 + i * 2]     = (v << 7) | (g << 4) | ((code >> 8) & 0x0F) as u8;
            buf[1 + i * 2 + 1] = (code & 0xFF) as u8;
        }
        self.inner.i2c.write(self.inner.addr, &buf)
    }

    /// Set V_REF for all four channels (volatile register only).
    pub fn set_vref(&mut self, vref_a: u8, vref_b: u8, vref_c: u8, vref_d: u8) -> Result<(), I2C::Error> {
        let byte1 = CMD_WRITE_VREF
            | ((vref_a & 0x01) << 3) | ((vref_b & 0x01) << 2)
            | ((vref_c & 0x01) << 1) |  (vref_d & 0x01);
        self.inner.i2c.write(self.inner.addr, &[byte1])
    }

    /// Set gain for all four channels (volatile register only).
    pub fn set_gain(&mut self, gain_a: u8, gain_b: u8, gain_c: u8, gain_d: u8) -> Result<(), I2C::Error> {
        let byte1 = CMD_WRITE_GAIN
            | ((if gain_a == 2 { 1 } else { 0 }) << 3)
            | ((if gain_b == 2 { 1 } else { 0 }) << 2)
            | ((if gain_c == 2 { 1 } else { 0 }) << 1)
            |  (if gain_d == 2 { 1 } else { 0 });
        self.inner.i2c.write(self.inner.addr, &[byte1])
    }

    /// Set power-down mode for all four channels (volatile register only).
    pub fn set_power_down(&mut self, pd_a: u8, pd_b: u8, pd_c: u8, pd_d: u8) -> Result<(), I2C::Error> {
        let byte1 = CMD_WRITE_POWERDOWN
            | (((pd_a >> 1) & 0x01) << 4) | ((pd_a & 0x01) << 3)
            | (((pd_b >> 1) & 0x01) << 2) | ((pd_b & 0x01) << 1);
        let byte2 = (((pd_c >> 1) & 0x01) << 6) | ((pd_c & 0x01) << 5)
                  | (((pd_d >> 1) & 0x01) << 4) | ((pd_d & 0x01) << 3);
        self.inner.i2c.write(self.inner.addr, &[byte1, byte2])
    }

    /// Read all four channels' DAC input registers and EEPROM contents.
    pub fn read(&mut self) -> Result<ReadResult, I2C::Error> {
        let mut buf = [0u8; 24];
        self.inner.i2c.read(self.inner.addr, &mut buf)?;
        let mut result = ReadResult {
            channel: [ChannelState { code: 0, vref: 0, gain: 1, power_down: 0,
                                    eeprom_code: 0, eeprom_vref: 0, eeprom_gain: 1, eeprom_power_down: 0 }; 4],
            eeprom_ready: (buf[0] & 0x80) != 0,
        };
        for i in 0..4 {
            let b = i * 3;
            result.channel[i].code       = (((buf[b + 1] & 0x0F) as u16) << 8) | buf[b + 2] as u16;
            result.channel[i].vref       = (buf[b + 1] >> 7) & 0x01;
            result.channel[i].power_down = (buf[b + 1] >> 5) & 0x03;
            result.channel[i].gain       = if (buf[b + 1] >> 4) & 0x01 != 0 { 2 } else { 1 };
        }
        for i in 0..4 {
            let b = 12 + i * 3;
            result.channel[i].eeprom_code       = (((buf[b + 1] & 0x0F) as u16) << 8) | buf[b + 2] as u16;
            result.channel[i].eeprom_vref       = (buf[b + 1] >> 7) & 0x01;
            result.channel[i].eeprom_power_down = (buf[b + 1] >> 5) & 0x03;
            result.channel[i].eeprom_gain       = if (buf[b + 1] >> 4) & 0x01 != 0 { 2 } else { 1 };
        }
        Ok(result)
    }

    /// Check if the EEPROM write is complete (RDY/BSY = 1).
    pub fn is_eeprom_ready(&mut self) -> Result<bool, I2C::Error> {
        let mut buf = [0u8; 1];
        self.inner.i2c.read(self.inner.addr, &mut buf)?;
        Ok((buf[0] & 0x80) != 0)
    }

    /// Send General Call Software Update (0x00, 0x08) to latch all V_OUT.
    pub fn software_update(&mut self) -> Result<(), I2C::Error> {
        self.inner.i2c.write(ADDR_GENERAL_CALL, &[GC_SOFTWARE_UPD])
    }

    /// Send General Call Wake-Up (0x00, 0x09) to clear all PD bits.
    pub fn wake_up(&mut self) -> Result<(), I2C::Error> {
        self.inner.i2c.write(ADDR_GENERAL_CALL, &[GC_WAKE])
    }

    /// Send General Call Reset (0x00, 0x06) to reload EEPROM into all DAC registers.
    pub fn reset(&mut self) -> Result<(), I2C::Error> {
        self.inner.i2c.write(ADDR_GENERAL_CALL, &[GC_RESET])
    }

    fn _single_write(&mut self, channel: u8, code: u16, vref: u8, pd: u8, gain: u8, udac: u8) -> Result<(), I2C::Error> {
        let ch = channel.min(3);
        let c = code.min(4095);
        let buf = [
            CMD_SINGLE_WRITE | ((ch & 0x03) << 1) | (udac & 0x01),
            ((vref & 0x01) << 7) | ((pd & 0x03) << 5) | ((gain & 0x01) << 4) | ((c >> 8) & 0x0F) as u8,
            (c & 0xFF) as u8,
        ];
        self.inner.i2c.write(self.inner.addr, &buf)
    }
}
